src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -973,11 +974,41 @@ public ICollection CustomMetadataNames
             /// <summary>
             /// Gets the number of custom metadata set on the item.
             /// Does not include built-in metadata.
-            /// Computed, not necessarily fast.
+            /// This is the count of <see cref="MetadataCollection"/>.
             /// </summary>
             public int CustomMetadataCount
             {
-                get { return CustomMetadataNames.Count; }
+                get
+                {
+                    int count = 0;
+
+                    // NOTE we walk collections here and exclude null values from counts to maintain parity
+                    // with the implementation of CustomMetadata.
+
+                    if (_itemDefinitions is not null)
+                    {
+                        foreach (ProjectItemDefinitionInstance item in _itemDefinitions)
+                        {
+                            if (item.Metadata is not null)
```

Comment:
I'm looking at this again and wondering if this is right. So it looks like the metadata should never be null—although there is a check like this in MetadataCollection, the `else` condition is just Debug.Fail, which suggests to me that's just an error condition. On the other hand, it uses allMetadata.Set to "add" the metadata because "Front of the list is highest priority," which to me suggests there might be duplicate metadata that should not be used. The same applies for the _directMetadata computation below.

I'm wondering if it might make sense to store the "current known" state of metadata separately from any recent changes. That way you could skip most of the work. What do you think?

Reply:
I would have liked to avoid these null checks here, and the enumeration, and just added `Count`, but that would have been a breaking behaviour change. If you are sure there's no negative consequence from breaking that behaviour and overreporting the count when nulls exist, we can change this. I'm not confident that it's safe to do, however.

Reply:
I traced it back _itemDefinitions back for a while and haven't found any way to add a null to it yet, though I didn't go all the way. The other comments about Set potentially _not_ increasing the count was the more pertinent one in my mind—that means this could create incorrect behavior in what seems to be an expected case.

Reply:
Discussion in https://github.com/dotnet/msbuild/issues/5267 suggests that the presence of null values may be due to concurrency issues with a prior implementation that predated the use of immutable collections.

Reply:
The important problem here is that in the old way for computing CustomMetadataNames, we looped through _itemDefinitions _and_, under that, we looped through individual metadata. Further, there can be duplicates between ProjectItemDefinitionInstances, so just counting how many there are might get you the wrong answer. I can't think of a way to avoid creating at least a set in calculating CustomMetadataCount unless it was already created, which gets back to:
> I'm wondering if it might make sense to store the "current known" state of metadata separately from any recent changes. That way you could skip most of the work.

The current way this is written will give incorrect metadata counts. We should also add a unit test to verify we don't hit this in the future.

Reply:
That's an awesome observation! I'll also note that grepping for `CustomMetadataCount` yields no hits in any relevant .NET or VS related repo and there's only one internal use in `TaskItem.Equals`. So one possible way of addressing may be leaving `CustomMetadataCount` as is and coming up with a better implementation of `Equals`.

Reply:
Thank you, that's a good point and invalidates the approach here as it is.

I've rolled back much of this PR locally, added a test for the previous behaviour, and am taking another look at this.